File: arduino/Incremental/colors.cpp
#include "colors.h"
#include "Arduino.h"

double unitSineWave(double x){
  // 0 to 1 to 0 along sine curve
  return 1 - 0.5 * (cos(2*PI*x) + 1);
}


uint8_t TransitionColorPart(uint8_t from, uint8_t to, double percent){
  return from + (to - from) * percent;
}
uint32_t TransitionColor(uint32_t from, uint32_t to, double percent){
  uint8_t r = TransitionColorPart(RedPart(from), RedPart(to), percent);
  uint8_t g = TransitionColorPart(GreenPart(from), GreenPart(to), percent);
  uint8_t b = TransitionColorPart(BluePart(from), BluePart(to), percent);
  uint8_t w = TransitionColorPart(WhitePart(from), WhitePart(to), percent);
  return AsColor(r, g, b, w);
}

uint32_t AsColor(uint8_t r, uint8_t g, uint8_t b, uint8_t w){
  return ((uint32_t)w << 24) | ((uint32_t)r << 16) | ((uint32_t)g <<  8) | b;
}
uint32_t AsColor(uint8_t r, uint8_t g, uint8_t b){
  return ((uint32_t)r << 16) | ((uint32_t)g <<  8) | b;
}
uint8_t WhitePart(uint32_t color){
  return (uint8_t)(color >> 24);
}
uint8_t RedPart(uint32_t color){
  return (uint8_t)(color >> 16);
}
uint8_t GreenPart(uint32_t color){
  return (uint8_t)(color >> 8);
}
uint8_t BluePart(uint32_t color){
  return (uint8_t)color;
}

uint32_t SetWhitePart(uint32_t color, uint8_t value){
   return ((uint32_t)value << 24) | ((uint32_t)(color >> 16) << 16) | ((uint32_t)(color >> 8) <<  8) | (color >> 16);
}
uint32_t SetRedPart(uint32_t color, uint8_t value){
   return ((uint32_t)(color >> 24) << 24) | ((uint32_t)value << 16) | ((uint32_t)(color >> 8) <<  8) | (color >> 16);
}
uint32_t SetGreenPart(uint32_t color, uint8_t value){
   return ((uint32_t)(color >> 24) << 24) | ((uint32_t)(color >> 16) << 16) | ((uint32_t)value <<  8) | (color >> 16);
}
uint32_t SetBluePart(uint32_t color, uint8_t value){
   return ((uint32_t)(color >> 24) << 24) | ((uint32_t)(color >> 16) << 16) | ((uint32_t)(color >> 8) <<  8) | value;
}
String ColorAsHex(uint32_t color){
  String rtrn = String(color, HEX);
  rtrn.toUpperCase();
  while (rtrn.length() < 8)
  { rtrn = "0" + rtrn; }
  return rtrn;
}

File: testPlatform/WinFormsApp1/Models/Colors.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WinFormsApp1.Models
{
    public static class Colors
    {
        public static ulong AsColor(byte r, byte g, byte b, byte w) => ((ulong)w << 24) | ((ulong)r << 16) | ((ulong)g << 8) | b;
        public static ulong AsColor(byte r, byte g, byte b) => ((ulong)r << 16) | ((ulong)g << 8) | b;
        public static byte WhitePart(ulong color) => (byte)(color >> 24);
        public static byte RedPart(ulong color) => (byte)(color >> 16);
        public static byte GreenPart(ulong color) => (byte) (color >> 8);
        public static byte BluePart(ulong color) => (byte)color;
        public static ulong SetWhitePart(ulong color, byte value) => ((ulong)value << 24) | ((ulong)(color >> 16) << 16) | ((ulong)(color >> 8) << 8) | (color >> 16);
        public static ulong SetRedPart(ulong color, byte value) => ((ulong)(color >> 24) << 24) | ((ulong)value << 16) | ((ulong)(color >> 8) << 8) | (color >> 16);
        public static ulong SetGreenPart(ulong color, byte value) => ((ulong)(color >> 24) << 24) | ((ulong)(color >> 16) << 16) | ((ulong)value << 8) | (color >> 16);
        public static ulong SetBluePart(ulong color, byte value) => ((ulong)(color >> 24) << 24) | ((ulong)(color >> 16) << 16) | ((ulong)(color >> 8) << 8) | value;
        public static ulong TransitionColor(ulong from, ulong to, double percent)
        {
            byte r = TransitionColorPart(RedPart(from), RedPart(to), percent);
            byte g = TransitionColorPart(GreenPart(from), GreenPart(to), percent);
            byte b = TransitionColorPart(BluePart(from), BluePart(to), percent);
            byte w = TransitionColorPart(WhitePart(from), WhitePart(to), percent);
            return AsColor(r, g, b, w);
        }
        public static string ColorAsHex(ulong color)
        {
            string rtrn = string.Format("{0:X}", color);
            while (rtrn.Length < 8)
            { rtrn = "0" + rtrn; }
            return rtrn;
            /*
            String rtrn = String(color, HEX);
            rtrn.toUpperCase();
            while (rtrn.length() < 8)
            { rtrn = "0" + rtrn; }
            return rtrn;*/
        }
        private static byte TransitionColorPart(byte from, byte to, double percent)
        {
            return (byte)(from + (to - from) * percent);
        }
    }
}
